compotime/models.py

```diff
@@ -585,12 +606,70 @@ def _log_mle_gen_var(X_zero: np.ndarray, g: np.ndarray, y: np.ndarray) -> float:
     """
     n = len(y)
     _, _, errors = _forward(X_zero, g, y)
-    _, gen_var_log = np.linalg.slogdet(
-        sum(error.reshape(-1, 1) @ error.reshape(1, -1) for error in errors) / n,
-    )
+
+    if np.isnan(y).any():
+        gen_var_log = _adj_log_mle_gen_var(y, errors)
+    else:
+        _, gen_var_log = np.linalg.slogdet(
+            sum(error.reshape(-1, 1) @ error.reshape(1, -1) for error in errors) / n,
+        )
     return gen_var_log
 
 
+def _adj_log_mle_gen_var(y: np.ndarray, errors: np.ndarray) -> float:
+    """Compute the logarithm of the adjusted maximum likelihood estimator for the generalized
+    variance for cases with time series of different lenghts (i.e. containing NaN values).
+
+    Parameters
+    ----------
+        y: Observed time series.
+        errors: Array of errors per timestamp.
+
+    Returns
+    -------
+        Logarithm of the adjusted maximum likelihood estimator for the generalized variance 
+        for TS with NaN values.
+    """
+    num_nan = (~np.isnan(y)).sum(axis=0)
+    V = np.zeros((y.shape[1], y.shape[1]))
+    for i in range(y.shape[1]):
+        for j in range(i, y.shape[1]):
+            V[i, j] = (errors[:, i] * errors[:, j]).sum() / min(num_nan[i], num_nan[j])
+            if i != j:
+                V[j, i] = V[i, j]
+    
+    adj_gen_var = 0
+    for y_t in y:
+        D = _select_matrix(y_t)
+        V_tilde = D @ V @ D.T
+        _, gen_var_log_t = np.linalg.slogdet(V_tilde)
+        adj_gen_var += gen_var_log_t
+    
+    return adj_gen_var
+
+
+def _select_matrix(y_t: np.ndarray) -> np.ndarray:
+    """Compute matrix that selects the series that are observed (different than 0) at time t.
+
+    Parameters
+    ----------
+        y_t: Values of each time series at time t.
+
+    Returns
+    -------
+        Selection matrix.
+    """
+    D = np.zeros((np.count_nonzero(~np.isnan(y_t)), len(y_t)))
```

Comment:
I'm not sure if I understand this. Why is the number of rows equal to the number non nan elements in the vector? Maybe I'm missing something, but shouldn't D be a diagonal (and hence square) matrix?

Can't D be computed  with just`D = np.diag(~np.isnan(y_t))`?

Reply:
As discussed, D is not a diagonal matrix and is correctly implemented